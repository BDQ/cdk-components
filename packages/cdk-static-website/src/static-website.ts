import * as path from 'path';
import {
  CfnCloudFrontOriginAccessIdentity,
  CloudFrontWebDistribution,
  CloudFrontWebDistributionProps,
  SSLMethod,
  SecurityPolicyProtocol
} from '@aws-cdk/aws-cloudfront';
import { AliasRecord, HostedZoneProvider } from '@aws-cdk/aws-route53';
import { Bucket } from '@aws-cdk/aws-s3';
import { BucketDeployment, Source } from '@aws-cdk/aws-s3-deployment';
import { Construct, SSMParameterProvider } from '@aws-cdk/cdk';
import { PolicyStatement } from '@aws-cdk/aws-iam';

import {
  LambdaFunctionAssociations,
  Association
} from './lambda-function-associations';

export interface AliasConfiguration {
  domainName: string;
  siteSubDomain: string;
}

export interface StaticWebsiteProps {
  /** Logical name of the site */
  siteName: string;

  aliasConfiguration?: AliasConfiguration;

  source?: string;

  disableUpload?: false;

  /** The index page for the site like 'index.html' */
  websiteIndexDocument?: string;

  /** The error page for the site like 'error.html' */
  websiteErrorDocument?: string;

  /** AWS WAF WebACL to associate with this CloudFront distribution */
  webACLId?: string;
}

export class StaticWebsite extends Construct {
  /** The static websites URL */
  public readonly url: string;

  private readonly distribution: CloudFrontWebDistribution;

  constructor(parent: Construct, name: string, props: StaticWebsiteProps) {
    super(parent, name);

    const {
      aliasConfiguration,
      siteName,
      disableUpload,
      source,
      websiteIndexDocument,
      websiteErrorDocument,
      webACLId
    } = props;

    const originId = new CfnCloudFrontOriginAccessIdentity(
      this,
      'OriginAccessIdentity',
      {
        cloudFrontOriginAccessIdentityConfig: {
          comment: `CloudFront OriginAccessIdentity for ${siteName}`
        }
      }
    );

    const websiteBucket = new Bucket(this, 'WebsiteBucket', {
      bucketName: siteName,
      websiteIndexDocument: websiteIndexDocument || 'index.html',
      websiteErrorDocument: websiteErrorDocument || 'error.html'
    });

    websiteBucket.addToResourcePolicy(
      new PolicyStatement()
        .addAction('s3:GetObject')
        .addCanonicalUserPrincipal(
          originId.cloudFrontOriginAccessIdentityS3CanonicalUserId
        )
        .addResource(`${websiteBucket.bucketArn}/*`)
    );

    if (!disableUpload) {
      const placeHolderSource = path.join(__dirname, '..', 'website');

      new BucketDeployment(this, 'DeployWebsite', {
        source: Source.asset(source || placeHolderSource),
        destinationBucket: websiteBucket
      });
    }

    const distibutionConfig: CloudFrontWebDistributionProps = {
      originConfigs: [
        {
          s3OriginSource: {
            originAccessIdentity: originId,
            s3BucketSource: websiteBucket
          },
          behaviors: [{ isDefaultBehavior: true }]
        }
      ]
    };

    this.distribution = aliasConfiguration
      ? this.configureDistributionWithAlias(
          aliasConfiguration,
          distibutionConfig
        )
      : new CloudFrontWebDistribution(
          this,
          'WebSiteDistribution',
          distibutionConfig
        );

    this.url = aliasConfiguration
      ? this.createSiteDomain(aliasConfiguration)
      : this.distribution.domainName;
  }

  public addLambdaFunctionAssociation(assosiation: Association) {
    return this.addLambdaFunctionAssociations([assosiation]);
  }

  public addLambdaFunctionAssociations(assosiations: Association[]) {
    return new LambdaFunctionAssociations(this, 'LambdaFunctionAssociation', {
      distribution: this.distribution,
      assosiations
    });
  }

  private createSiteDomain(aliasConfiguration: AliasConfiguration) {
    const { siteSubDomain, domainName } = aliasConfiguration;

    return `${siteSubDomain}.${domainName}`;
  }

  private configureDistributionWithAlias(
    aliasConfiguration: AliasConfiguration,
    distibutionConfig: CloudFrontWebDistributionProps
  ) {
    const siteDomain = this.createSiteDomain(aliasConfiguration);

    const certificateArn = new SSMParameterProvider(this, {
      parameterName: 'CertificateArn-' + siteDomain
    }).parameterValue();

    if (certificateArn === 'dummy') {
      throw new Error(`No parameter with name CertificateArn-${siteDomain}`);
    }

    distibutionConfig.aliasConfiguration = {
      acmCertRef: certificateArn,
      names: [siteDomain],
      sslMethod: SSLMethod.SNI,
      securityPolicy: SecurityPolicyProtocol.TLSv1_1_2016
    };

    const distribution = new CloudFrontWebDistribution(
      this,
      'WebSiteDistribution',
      distibutionConfig
    );

    const { domainName } = aliasConfiguration;

    const zone = new HostedZoneProvider(this, {
      domainName
    }).findAndImport(this, 'Zone');

    new AliasRecord(this, 'WebsiteAliasRecord', {
      zone,
      recordName: siteDomain,
      target: distribution
    });

    return distribution;
  }
}
