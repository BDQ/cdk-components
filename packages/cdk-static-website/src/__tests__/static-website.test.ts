import { Stack } from '@aws-cdk/core';
import 'jest-cdk-snapshot';

import { StaticWebsite } from '../static-website';

test('default setup', (): void => {
    const stack = new Stack();

    new StaticWebsite(stack, 'StaticWebsite', {
        disableUpload: true,
    });

    expect(stack).toMatchCdkSnapshot();
});
