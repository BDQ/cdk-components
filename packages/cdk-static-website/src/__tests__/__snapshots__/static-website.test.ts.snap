// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`default setup 1`] = `
Object {
  "Parameters": Object {
    "AssetParameters706389260489e05ba948c7d1a6d6aaf58792b1e033956edc0bf0e530820cae75ArtifactHashA3EE1CE3": Object {
      "Description": "Artifact hash for asset \\"706389260489e05ba948c7d1a6d6aaf58792b1e033956edc0bf0e530820cae75\\"",
      "Type": "String",
    },
    "AssetParameters706389260489e05ba948c7d1a6d6aaf58792b1e033956edc0bf0e530820cae75S3Bucket81B1BD07": Object {
      "Description": "S3 bucket for asset \\"706389260489e05ba948c7d1a6d6aaf58792b1e033956edc0bf0e530820cae75\\"",
      "Type": "String",
    },
    "AssetParameters706389260489e05ba948c7d1a6d6aaf58792b1e033956edc0bf0e530820cae75S3VersionKeyFBE89C0F": Object {
      "Description": "S3 key for asset version \\"706389260489e05ba948c7d1a6d6aaf58792b1e033956edc0bf0e530820cae75\\"",
      "Type": "String",
    },
    "AssetParametersa9125fa9a40550c71cde90bd478cc23091e868067a12380c1df0827d013ad2ffArtifactHash08605F5E": Object {
      "Description": "Artifact hash for asset \\"a9125fa9a40550c71cde90bd478cc23091e868067a12380c1df0827d013ad2ff\\"",
      "Type": "String",
    },
    "AssetParametersa9125fa9a40550c71cde90bd478cc23091e868067a12380c1df0827d013ad2ffS3Bucket848A1F31": Object {
      "Description": "S3 bucket for asset \\"a9125fa9a40550c71cde90bd478cc23091e868067a12380c1df0827d013ad2ff\\"",
      "Type": "String",
    },
    "AssetParametersa9125fa9a40550c71cde90bd478cc23091e868067a12380c1df0827d013ad2ffS3VersionKey983DBE96": Object {
      "Description": "S3 key for asset version \\"a9125fa9a40550c71cde90bd478cc23091e868067a12380c1df0827d013ad2ff\\"",
      "Type": "String",
    },
  },
  "Resources": Object {
    "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756C81C01536": Object {
      "DependsOn": Array [
        "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRoleDefaultPolicy88902FDF",
        "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRole89A01265",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParametersa9125fa9a40550c71cde90bd478cc23091e868067a12380c1df0827d013ad2ffS3Bucket848A1F31",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParametersa9125fa9a40550c71cde90bd478cc23091e868067a12380c1df0827d013ad2ffS3VersionKey983DBE96",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParametersa9125fa9a40550c71cde90bd478cc23091e868067a12380c1df0827d013ad2ffS3VersionKey983DBE96",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRole89A01265",
            "Arn",
          ],
        },
        "Runtime": "python3.6",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRole89A01265": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRoleDefaultPolicy88902FDF": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      Object {
                        "Ref": "AssetParameters706389260489e05ba948c7d1a6d6aaf58792b1e033956edc0bf0e530820cae75S3Bucket81B1BD07",
                      },
                    ],
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      Object {
                        "Ref": "AssetParameters706389260489e05ba948c7d1a6d6aaf58792b1e033956edc0bf0e530820cae75S3Bucket81B1BD07",
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
            Object {
              "Action": Array [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
                "s3:DeleteObject*",
                "s3:PutObject*",
                "s3:Abort*",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "StaticWebsiteBucketWebsiteBucket9EE4F45B",
                    "Arn",
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      Object {
                        "Fn::GetAtt": Array [
                          "StaticWebsiteBucketWebsiteBucket9EE4F45B",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRoleDefaultPolicy88902FDF",
        "Roles": Array [
          Object {
            "Ref": "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRole89A01265",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "StaticWebsiteBucketDeploymentCustomResourceBC04F1EE": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "DestinationBucketName": Object {
          "Ref": "StaticWebsiteBucketWebsiteBucket9EE4F45B",
        },
        "RetainOnDelete": true,
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756C81C01536",
            "Arn",
          ],
        },
        "SourceBucketNames": Array [
          Object {
            "Ref": "AssetParameters706389260489e05ba948c7d1a6d6aaf58792b1e033956edc0bf0e530820cae75S3Bucket81B1BD07",
          },
        ],
        "SourceObjectKeys": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters706389260489e05ba948c7d1a6d6aaf58792b1e033956edc0bf0e530820cae75S3VersionKeyFBE89C0F",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters706389260489e05ba948c7d1a6d6aaf58792b1e033956edc0bf0e530820cae75S3VersionKeyFBE89C0F",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        ],
      },
      "Type": "Custom::CDKBucketDeployment",
      "UpdateReplacePolicy": "Delete",
    },
    "StaticWebsiteBucketOriginAccessIdentity7BF15D42": Object {
      "Properties": Object {
        "CloudFrontOriginAccessIdentityConfig": Object {
          "Comment": Object {
            "Fn::Join": Array [
              "",
              Array [
                "CloudFront OriginAccessIdentity for ",
                Object {
                  "Ref": "StaticWebsiteBucketWebsiteBucket9EE4F45B",
                },
              ],
            ],
          },
        },
      },
      "Type": "AWS::CloudFront::CloudFrontOriginAccessIdentity",
    },
    "StaticWebsiteBucketWebsiteBucket9EE4F45B": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "WebsiteConfiguration": Object {
          "ErrorDocument": "error.html",
          "IndexDocument": "index.html",
        },
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "StaticWebsiteBucketWebsiteBucketPolicyD8AEC0D4": Object {
      "Properties": Object {
        "Bucket": Object {
          "Ref": "StaticWebsiteBucketWebsiteBucket9EE4F45B",
        },
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Principal": Object {
                "CanonicalUser": Object {
                  "Fn::GetAtt": Array [
                    "StaticWebsiteBucketOriginAccessIdentity7BF15D42",
                    "S3CanonicalUserId",
                  ],
                },
              },
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "StaticWebsiteBucketWebsiteBucket9EE4F45B",
                    "Arn",
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      Object {
                        "Fn::GetAtt": Array [
                          "StaticWebsiteBucketWebsiteBucket9EE4F45B",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::S3::BucketPolicy",
    },
    "StaticWebsiteDistributionCFDistribution4F2E76E7": Object {
      "Properties": Object {
        "DistributionConfig": Object {
          "DefaultCacheBehavior": Object {
            "AllowedMethods": Array [
              "GET",
              "HEAD",
            ],
            "CachedMethods": Array [
              "GET",
              "HEAD",
            ],
            "Compress": true,
            "ForwardedValues": Object {
              "Cookies": Object {
                "Forward": "none",
              },
              "QueryString": false,
            },
            "TargetOriginId": "origin1",
            "ViewerProtocolPolicy": "redirect-to-https",
          },
          "DefaultRootObject": "index.html",
          "Enabled": true,
          "HttpVersion": "http2",
          "IPV6Enabled": true,
          "Origins": Array [
            Object {
              "DomainName": Object {
                "Fn::GetAtt": Array [
                  "StaticWebsiteBucketWebsiteBucket9EE4F45B",
                  "RegionalDomainName",
                ],
              },
              "Id": "origin1",
              "S3OriginConfig": Object {
                "OriginAccessIdentity": Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "origin-access-identity/cloudfront/",
                      Object {
                        "Ref": "StaticWebsiteBucketOriginAccessIdentity7BF15D42",
                      },
                    ],
                  ],
                },
              },
            },
          ],
          "PriceClass": "PriceClass_100",
          "ViewerCertificate": Object {
            "CloudFrontDefaultCertificate": true,
          },
        },
      },
      "Type": "AWS::CloudFront::Distribution",
    },
  },
}
`;
